source/opt/eliminate_dead_functions_pass.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "eliminate_dead_functions_pass.h"
+
+#include <unordered_set>
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status EliminateDeadFunctionsPass::Process(ir::Module* module) {
+  bool modified = false;
+
+  // Identify live functions first.  Those that are not live
+  // are dead.
+  std::unordered_set<const ir::Function*> liveFunctionSet;
+  ProcessFunction pfn = [&liveFunctionSet](ir::Function* fp) {
+    liveFunctionSet.insert(fp);
+    return false;
+  };
+  ProcessReachableCallTree(pfn, module);
+
+  def_use_mgr_.reset(new analysis::DefUseManager(consumer(), module));
+  for (auto funcIter = module->begin(); funcIter != module->end();) {
+    if (liveFunctionSet.count(&*funcIter) == 0) {
+      modified = true;
+      EliminateFunction(&*funcIter);
+      funcIter = funcIter.Erase();
+    } else {
+      ++funcIter;
+    }
+  }
+
+  return modified ? Pass::Status::SuccessWithChange
+                  : Pass::Status::SuccessWithoutChange;
+}
+
+void EliminateDeadFunctionsPass::EliminateFunction(ir::Function* func) {
+  // Remove all of the instruction in the function body
+  func->ForEachInst(
```

Comment:
It's kind of awesome it boils down to just this.

Reply:
Yes, and the Erase on the function iterator is pretty cool too!